include/spdk/nvme.h

```diff
@@ -1141,6 +1142,36 @@ int spdk_nvme_ns_cmd_reservation_report(struct spdk_nvme_ns *ns,
 					void *payload, uint32_t len,
 					spdk_nvme_cmd_cb cb_fn, void *cb_arg);
 
+/**
+ * \brief Sends an abort command to the Nvme device for the request indicated
+ *  by cid and sqid
+ *
+ * \param ctrlr controller to the Nvme device
+ * \param cid identification number of the command to be aborted.
+ * \param sqid submission queue number of the command to be aborted
+ * \param cb_fn a function pointer to the abort call back function
+ * \param cb_arg user argument to the abort command callback function
+ *
+ * \return 0 if successfully submitted, Less than 0 if it fails
+ *
+ * Note: The user must ensure to decrement the controller member
```

Comment:
The timeout callback will always have the ctrlr_lock already held, since it is called in response to a spdk_nvme_process_admin_completions() function call, so the timeout completion handler code can assume it already has the lock.

It might be simplest to just not add the abort count limit checking in this first version; the worst that can happen is that the Abort command will be failed by the controller with Abort Command Limit Exceeded status, and then the abort completion handler can immediately issue a controller reset to recover.  (If enough commands time out simultaneously that the abort limit is exceeded, the controller is probably not responsive anyway.)

If we do want to enforce the abort limit, then I still think it should be entirely handled inside the NVMe library.  The simplest way to do this would be to have an intermediate callback function that does the decrement internally.  An example where the NVMe library already does this sort of intermediate callback is the nvme_user_copy_cmd_complete() path; you can reuse the existing user_cb_fn and user_cb_arg fields to store the actual user abort callback.